### ISSUE TITLE
`pnpm rebuild` fails

### ISSUE THREAD

**Issue by vjpr** [pnpm/pnpm#756]
When I run `pnpm rebuild` or `npm rebuild`, it fails to run. During a regular install it works though.

It seems like when `npm rebuild` is run, it does not resolve modules to their realpath. Therefore it is searching in `app/node_modules/.bin/` to find `node-pre-gyp`, instead of `app/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules/.bin/node-pre-gyp`.

```
$ npm rebuild

> bcrypt@1.0.2 install /Users/Vaughan/foo/app--worktree--master/node_modules/bcrypt
> node-pre-gyp install --fallback-to-build

module.js:457
    throw err;
    ^

Error: Cannot find module '/Users/Vaughan/foo/app--worktree--master/node_modules/node-pre-gyp/bin/node-pre-gyp'
    at Function.Module._resolveFilename (module.js:455:15)
    at Function.Module._load (module.js:403:25)
    at Module.runMain (module.js:590:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3

npm ERR! Darwin 16.5.0
npm ERR! argv "/Users/Vaughan/nvm/versions/node/v6.5.0/bin/node" "/Users/Vaughan/nvm/versions/node/v6.5.0/bin/npm" "rebuild"
npm ERR! node v6.5.0
npm ERR! npm  v4.2.0
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! bcrypt@1.0.2 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the bcrypt@1.0.2 install script 'node-pre-gyp install --fallback-to-build'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the bcrypt package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-pre-gyp install --fallback-to-build
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs bcrypt
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls bcrypt
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/Vaughan/.npm/_logs/2017-05-08T11_58_09_693Z-debug.log
```

Extract from `pnpm install` scripts stage logs.

```
...
 registry.npmjs.org/bcrypt/1.0.2  $ npm run install
  registry.npmjs.org/bcrypt/1.0.2  > bcrypt@1.0.2 install /Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules/bcrypt
  registry.npmjs.org/bcrypt/1.0.2  > node-pre-gyp install --fallback-to-build
  registry.npmjs.org/bcrypt/1.0.2! node-pre-gyp ERR! Tried to download(404): https://github.com/kelektiv/node.bcrypt.js/releases/download/v1.0.2/bcrypt_lib-v1.0.2-node-v48-darwin-x64.tar.gz
  registry.npmjs.org/bcrypt/1.0.2! node-pre-gyp ERR! Pre-built binaries not found for bcrypt@1.0.2 and node@6.5.0 (node-v48 ABI) (falling back to source compile with node-gyp)
  registry.npmjs.org/node-sass/3.8.0  Binary downloaded and installed at /Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-sass/3.8.0/node_modules/node-sass/vendor/darwin-x64-48/binding.node
  registry.npmjs.org/node-sass/3.8.0  > node-sass@3.8.0 postinstall /Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-sass/3.8.0/node_modules/node-sass
  registry.npmjs.org/node-sass/3.8.0  > node scripts/build.js
  registry.npmjs.org/ursa/0.9.4! gyp info spawn make
  registry.npmjs.org/ursa/0.9.4! gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  registry.npmjs.org/ursa/0.9.4    CXX(target) Release/obj.target/ursaNative/src/ursaNative.o
  registry.npmjs.org/node-sass/3.8.0  "/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-sass/3.8.0/node_modules/node-sass/vendor/darwin-x64-48/binding.node" exists.
  registry.npmjs.org/node-sass/3.8.0   testing binary.
  registry.npmjs.org/node-sass/3.8.0  Binary is fine; exiting.
  registry.npmjs.org/jsonpath/0.2.2  $ npm run postinstall
  registry.npmjs.org/jsonpath/0.2.2  > jsonpath@0.2.2 postinstall /Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/jsonpath/0.2.2/node_modules/jsonpath
  registry.npmjs.org/jsonpath/0.2.2  > node lib/aesprim.js > generated/aesprim-browser.js
  registry.npmjs.org/deasync/0.1.9  $ npm run install
  registry.npmjs.org/bcrypt/1.0.2    CXX(target) Release/obj.target/bcrypt_lib/src/blowfish.o
  registry.npmjs.org/deasync/0.1.9  > deasync@0.1.9 install /Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/deasync/0.1.9/node_modules/deasync
  registry.npmjs.org/deasync/0.1.9  > node ./build.js
  registry.npmjs.org/deasync/0.1.9  `darwin-x64-node-6` exists; testing
  registry.npmjs.org/bcrypt/1.0.2    CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt.o
  registry.npmjs.org/deasync/0.1.9  Binary is fine; exiting
  registry.npmjs.org/bufferutil/1.2.1  $ /Users/Vaughan/nvm/versions/node/v6.5.0/lib/node_modules/pnpm/lib/node_modules/node-gyp/bin/node-gyp.js rebuild
  registry.npmjs.org/bcrypt/1.0.2    CXX(target) Release/obj.target/bcrypt_lib/src/bcrypt_node.o
  registry.npmjs.org/bufferutil/1.2.1! gyp info it worked if it ends with ok
  registry.npmjs.org/bufferutil/1.2.1! gyp info using node-gyp@3.6.1
  registry.npmjs.org/bufferutil/1.2.1! gyp info using node@6.5.0 | darwin | x64
  registry.npmjs.org/ursa/0.9.4    SOLINK_MODULE(target) Release/ursaNative.node
  registry.npmjs.org/ursa/0.9.4! clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn /usr/local/bin/python2
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args [ '/Users/Vaughan/nvm/versions/node/v6.5.0/lib/node_modules/pnpm/lib/node_modules/node-gyp/gyp/gyp_main.py',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   'binding.gyp',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-f',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   'make',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-I',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bufferutil/1.2.1/node_modules/bufferutil/build/config.gypi',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-I',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '/Users/Vaughan/nvm/versions/node/v6.5.0/lib/node_modules/pnpm/lib/node_modules/node-gyp/addon.gypi',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-I',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '/Users/Vaughan/.node-gyp/6.5.0/include/node/common.gypi',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Dlibrary=shared_library',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Dvisibility=default',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Dnode_root_dir=/Users/Vaughan/.node-gyp/6.5.0',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Dnode_gyp_dir=/Users/Vaughan/nvm/versions/node/v6.5.0/lib/node_modules/pnpm/lib/node_modules/node-gyp',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Dnode_lib_file=node.lib',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Dmodule_root_dir=/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bufferutil/1.2.1/node_modules/bufferutil',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Dnode_engine=v8',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '--depth=.',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '--no-parallel',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '--generator-output',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   'build',
  registry.npmjs.org/bufferutil/1.2.1! gyp info spawn args   '-Goutput_dir=.' ]
  registry.npmjs.org/ursa/0.9.4! gyp info ok
  registry.npmjs.org/utf-8-validate/1.2.2  $ /Users/Vaughan/nvm/versions/node/v6.5.0/lib/node_modules/pnpm/lib/node_modules/node-gyp/bin/node-gyp.js rebuild
  registry.npmjs.org/utf-8-validate/1.2.2! gyp info it worked if it ends with ok
  registry.npmjs.org/utf-8-validate/1.2.2! gyp info using node-gyp@3.6.1
  registry.npmjs.org/utf-8-validate/1.2.2! gyp info using node@6.5.0 | darwin | x64
  registry.npmjs.org/bcrypt/1.0.2    SOLINK_MODULE(target) Release/bcrypt_lib.node
  registry.npmjs.org/bcrypt/1.0.2! clang: warning: libstdc++ is deprecated; move to libc++ with a minimum deployment target of OS X 10.9 [-Wdeprecated]
  registry.npmjs.org/bcrypt/1.0.2    COPY /Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules/bcrypt/lib/binding/bcrypt_lib.node
  registry.npmjs.org/bcrypt/1.0.2    TOUCH Release/obj.target/action_after_build.stamp
...
```
### Additional information:

 - `node -v` prints: 6.5.0
 - Windows, OS X, or Linux?: macOS

**Comment by vjpr** [comment]
Or perhaps it is the bin wrapper.

```
➜  .bin git:(master) ✗ pwd
/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/bcrypt/node_modules/.bin

➜  .bin git:(master) ✗ cat node-pre-gyp
#!/bin/sh
basedir=$(dirname "$(echo "$0" | sed -e 's,\\,/,g')")

case `uname` in
    *CYGWIN*) basedir=`cygpath -w "$basedir"`;;
esac

if [ -x "$basedir/node" ]; then
  NODE_PATH="/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-pre-gyp/0.6.32/node_modules/node-pre-gyp/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-pre-gyp/0.6.32/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-pre-gyp/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules:/Users/Vaughan/dev-quantitec/node_modules:/Users/Vaughan/node_modules:/Users/node_modules:/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules/node-pre-gyp/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/node_modules" "$basedir/node"  "$basedir/../../../node-pre-gyp/bin/node-pre-gyp" "$@"
  ret=$?
else
  NODE_PATH="/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-pre-gyp/0.6.32/node_modules/node-pre-gyp/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-pre-gyp/0.6.32/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node-pre-gyp/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules:/Users/Vaughan/dev-quantitec/node_modules:/Users/Vaughan/node_modules:/Users/node_modules:/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules/node-pre-gyp/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules:/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/node_modules" node  "$basedir/../../../node-pre-gyp/bin/node-pre-gyp" "$@"
  ret=$?
fi
exit $ret

```

**Comment by vjpr** [comment]
So I get the same error when running the bin wrapper from the project root which must be what `npm rebuild` is doing.

```
➜  intranav-app--worktree--master git:(master) ✗ node_modules/bcrypt/node_modules/.bin/node-pre-gyp install
module.js:457
    throw err;
    ^

Error: Cannot find module '/Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/node-pre-gyp/bin/node-pre-gyp'
    at Function.Module._resolveFilename (module.js:455:15)
    at Function.Module._load (module.js:403:25)
    at Module.runMain (module.js:590:10)
    at run (bootstrap_node.js:394:7)
    at startup (bootstrap_node.js:149:9)
    at bootstrap_node.js:509:3
```

It looks like the pnpm bin wrappers assume they will be run with cwd as the package root.

**Comment by vjpr** [comment]
Actually looks like it is the realpath issue I first mentioned:

`pnpm install`:

```
  registry.npmjs.org/bcrypt/1.0.2  > bcrypt@1.0.2 install /Users/Vaughan/dev-quantitec/intranav-app--worktree--master/node_modules/.registry.npmjs.org/bcrypt/1.0.2/node_modules/bcrypt
```

`pnpm rebuild`:
```
> bcrypt@1.0.2 install /Users/Vaughan/foo/app--worktree--master/node_modules/bcrypt
```

**Comment by vjpr** [comment]
Ok looks like pnpm does it own native package building using `install/postInstall.ts`.

We should add `pnpm rebuild` as a supported command and run `postInstall`.

**Comment by vjpr** [comment]
Here is npm's rebuild: https://github.com/npm/npm/blob/master/lib/rebuild.js
and build: https://github.com/npm/npm/blob/master/lib/build.js
Docs: https://docs.npmjs.com/cli/build

I think we can scan all the deps in the shrinkwrap and pass them to `postInstall` can't we?

**Comment by zkochan** [comment]
Published in [v1.16.0](https://github.com/pnpm/pnpm/releases/tag/v1.16.0)
